sdks/python/apache_beam/io/textio.py

```diff
@@ -362,6 +391,15 @@ def _is_self_overlapping(delimiter):
         return True
     return False
 
+  def _is_escaped(self, read_buffer, position):
+    # Returns True if byte at position is preceded with an odd number
+    # of escapechar bytes or False if preceded by 0 or even escapes
+    # (the even number means that all the escapes are escaped themselves).
+    for current_pos in reversed(range(-1, position)):
+      if read_buffer.data[current_pos:current_pos + 1] != self._escapechar:
```

Comment:
This logic is a little tricky to follow:
1. why do we go to -1 ?
2. can we compare a single byte at a position instead of a 1-character range?

why not to count `preceding_escape_chars` in a loop and then  `return preceding_escape_chars % 2 == 1` ?

Reply:
Totally agree about the trickiness. But it is beautiful and works :) 
I will look for more explicit solution if you insist, though.

1. -1 is needed to check if the 1st character is escaped without additional check if it is the first or not and be able to return from the same if block if we reached the start of buffer, -1:0 range returns empty bytes which cannot be equal to escapechar as we do not allow empty one:
```
>>> b'\\'[-1:0]
b''
>>> b'\\'[-1:0] == b'\\'
False
```
2. No, we cannot compare `byte` to `bytes` (I was a bit surprised as well, but we are doing similar comparisons in `_find_separator_bounds`):
```
>>> b'a'[0] == b'a'
False
>>> b'a'[0:1] == b'a'
True
```

Reply:
It looks like the last line `return False` cannot be reached, because we loop until first non-escape or the start of buffer and always return. That's the problem...
It only passes Pylint validation because `inconsistent-return-statements` is disabled, but can confuse readers.

Reply:
Thanks. 
I personally like `explicit is better than implicit` from https://www.python.org/dev/peps/pep-0020/ . Someone reading this code might wonder whether [-1:0] range was an intentional choice to test implicitly test edge case, or it's an off-by-one error.

Reply:
I have updated the code to use an explicit counter. Now it is easier to understand.

Should I replace bytes slicing comparison with `if read_buffer.data[current_pos] != self._escapechar[0]`, or let's keep it consistent with `_find_separator_bounds`?

What about my suggestion in the first comment to have 2 `_find_separator_bounds` - one for the default case, another for custom delimiter and/or escapechar to avoid extra ifs in the default use case?

I can prepare such version to see how it would look, if you are ready to consider such code duplication for performance reasons.

Reply:
I doubt that there is a significant performance impact, but you could try to run https://github.com/apache/beam/blob/9a44c9e6e1d944cc022d3e3e11b7ec08f7a3a790/.test-infra/jenkins/job_PerformanceTests_Python.groovy#L39